Allow polygonal windows in AirflowNetwork

lgu1234 opened the pull request:
Pull request overview
---------------------
This new feature allows polygonal windows used in the AirflowNetwork model by converting a polygonal window into an equivalent rectangular shape.
### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [x] Verify naming conventions and styles, memos and notes and defaults
   - [x] Open windows IDF Editor with modified IDD to check for errors
   - [x] If transition, add rules to spreadsheet
   - [x] If transition, add transition source
   - [x] If transition, update idfs
 - [x] If new idf included, locally check the err file and other outputs
 - [x] Documentation changes in place
 - [x] Changed docs build successfully
 - [x] ExpandObjects changes?
 - [x] If output changes, including tabular output structure, add to output rules file for interfaces

lgu1234 committed:
Upload an NFP

Myoldmopar commented:
@lgu1234 I renamed the PR, feel free to change it to better reflect the new feature if needed.

lgu1234 commented:
@Myoldmopar Thanks for your renaming.

lgu1234 committed:
Update an NFP

The NFP was revised based on comments from Mike, Jason and Edwin.

jasondegraw commented on design/FY2016/NFP-AFNPolygonalWindows.md:
```diff
@@ -0,0 +1,491 @@
+\subsubsection{Field: Equivalent Rectangular Shape Choice}\label{equivalent-rectangular-shape- choice}bold red</span>**
+
+**<span style="color:red;">
+This field is applied to a non-rectangular window or door. The equivalent surface has the same area as the non-rectangular one. When Height is entered, the equivalent width is equal to the area divided by the height. When "Aspect Ratio from Input" is entered, the equivalent height is equal to Square Root of Area divided by AspectRatio. The equivalent width is equal to the equivalent height * Aspe ...
```
@lgu1234 I apologize if I'm just being dense, but where does the aspect ratio value come from if the "Aspect Ratio from Input" option is chosen?

lgu1234 replied:
@jasondegraw You are right. It should be from an additional field. I am going to add an optional field with default value at 1.

lgu1234 committed:
Added a new field for Aspect Ratio

jasondegraw committed:
Move the NFP to the FY2017 folder, delete png

mjwitte commented:
@lgu1234 @jasondegraw In general I'm OK with the concept, but I'd suggest some changes in the field names and choice keys.

```
A8, \field Equivalent Rectangle Aspect Ratio Method
  \type choice
  \key PolygonHeight
  \key ParentSurfaceAspectRatio
  \key UserDefined (or UserInput)
  \default PolygonHeight
  \note This field is applied to a non-rectangular window or door. The equivalent shape has
  \note the same area as a polygonal window or door. 
N7; \field Equivalent Rectangle Aspect Ratio 
  \note This field is used when UserDefined is entered in theEquivalent Rectangle Aspect Ratio Method field. 
  \units dimensionless 
  \type real 
  \minimum> 0.0 
  \default 1.0
```

lgu1234 commented:
@mjwitte I like your changes of choices. However, I may have a different opinion of the field name you suggested: Equivalent Rectangle Aspect Ratio Method. Two of 3 choices are aspect ratio related. However, the choice of PolygonHeight is not related to aspect ratio. I would like to make a change to: Equivalent Rectangle Method (by dropping Aspect Ratio). What do you think?

mjwitte commented:
I see. Yes, that works.

mjwitte commented:
But then "UserDefined" should now be "UserDefinedAspectRatio".

mjwitte commented:
And "ParentSurfaceAspectRatio" should be "BaseSurfaceAspectRatio" to be consistent with the surface input terminology.

lgu1234 commented:
@mjwitte Great. I like suggestions and will make the changes right now.

lgu1234 committed:
Revised field and choice names

The revision was based on Mike's comments.

lgu1234 committed:
Upload Design Document

lgu1234 committed:
Updated design document

The revision is based on Jason's comments

jasondegraw commented on design/FY2017/NFP-AFNPolygonalWindows.md:
```diff
@@ -0,0 +1,592 @@
+
+	int const EquivRec_Height( 1 ); // Effective rectangle polygonal height selection
+	int const EquivRec_BaseAspRatio( 2 ); // Effective rectangle base surface aspect ratio selection
+	int const EquivRec_UserAspRatio( 3 ); // Effective rectangle user input aspect ratio selection
```
@lgu1234 Can we not abbreviate "Aspect" in these names? The precedent in the rest of the code is to fully spell it out, and `EquivRec_BaseAspectRatio` and `EquivRec_UserAspectRatio` are a lot clearer to me.

lgu1234 replied:
@jasondegraw No problem. I am going to use  EquivRec_BaseAspectRatio  and  EquivRec_UserAspectRatio.

jasondegraw commented on design/FY2017/NFP-AFNPolygonalWindows.md:
```diff
@@ -0,0 +1,592 @@
+
+		bool NonRectangular( false ); // True if this surface is not rectangular
+		int EquivRecMethod( 1 ); // Equivalent Rectangle Method input: 1 Height; 2 Base surface aspect ratio; 3 User input aspect ratio
+		Real64 EquivRecUserAspRatio( 1.0 ); // user input value when EquivRecMethod = 3 
```
Same here, I'd rather it be `EquivRecUserAspectRatio`

lgu1234 replied:
@jasondegraw Sure. I ma going to make it as EquivRecUserAspectRatio.

lgu1234 committed:
Revised again by changing member names

lgu1234 committed:
Upload initial code and unit test

A rectangular window in AirflowNetwork3zVent.idf is splitted into two triangular windows. Differences are expected.

lgu1234 committed:
Merge branch 'develop' into AFNPolygonalWindowSupport

jasondegraw commented on testfiles/AirflowNetwork3zVent.idf:
```diff
@@ -743,9 +743,23 @@
+    3,                       !- Number of Vertices
     1.00000,0,2.500000,  !- X,Y,Z ==> Vertex 1 {m}
     1.00000,0,1.0000000,  !- X,Y,Z ==> Vertex 2 {m}
+    5.000000,0,1.0000000;  !- X,Y,Z ==> Vertex 3 {m}
```
@lgu1234 I tend to think that dividing a rectangular window into triangular windows is not something we should do in one of the existing example files. Could we either make a new example file or add some notes to indicate that we don't expect this to be needed in most cases?

Also, is this supposed to be `5.000000,0,2.500000;`?

lgu1234 replied:
@jasondegraw Thanks for finding a mistake. The correct coordinates have been changed. The window view is shown below.

![window](https://cloud.githubusercontent.com/assets/5970656/20971038/01ee5332-bc5e-11e6-8e12-3bfc007a0dc7.JPG)
 
I prefer to add notes, instead of adding one more example file, since we already have so many example files. Here are notes I would like to add under Highlights in the file:

A rectangular window was split into two triangular windows to demonstrate how the AirflowNetwrok model handles non-rectangular windows. Rectangular windows are recommended to be used in most cases.

Please let me know what you think.

jasondegraw replied:
@lgu1234 I think that would do it. Add that and then I'll run down the checklist and we can get this merged.

jasondegraw committed:
Minor documentation changes

jasondegraw committed:
Minor unit test changes

lgu1234 committed:
Updated an example file

lgu1234 committed:
Merge branch 'AFNPolygonalWindowSupport' of https://github.com/NREL/EnergyPlus into AFNPolygonalWindowSupport

jasondegraw commented:
@Myoldmopar Are you happy with the PR title? Everything looks good to me, so I'm probably going to merge this before the end of the day. So last chance to make a change...

Myoldmopar commented:
![image](https://cloud.githubusercontent.com/assets/849824/20988882/ddd91dce-bc8f-11e6-9540-61041bba2c2f.png)
